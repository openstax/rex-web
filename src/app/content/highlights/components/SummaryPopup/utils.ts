import { Highlight } from '@openstax/highlighter/dist/api';
import queryString from 'querystring';
import React from 'react';
import { useSelector } from 'react-redux';
import { systemQueryParameters } from '../../../../navigation/selectors';
import { Book } from '../../../types';
import { findArchiveTreeNodeById } from '../../../utils/archiveTreeUtils';
import { getBookPageUrlAndParams } from '../../../utils/urlUtils';

export const useCreateHighlightLink = ({ sourceId, id, anchor: elementId }: Highlight, book?: Book) => {
  const [link, setLink] = React.useState('');
  const systemQueryParams = useSelector(systemQueryParameters);

  React.useEffect(() => {
    const page = book ? findArchiveTreeNodeById(book.tree, sourceId) : undefined;
    const data: Record<string, string> = {
      ...systemQueryParams,
      target: JSON.stringify({ id, type: 'highlight' }),
    };
    const target = queryString.stringify(data) + `#${elementId}`;
    setLink(`${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${target}`);
  }, [id, book, elementId, sourceId, systemQueryParams]);

  return link;
};
