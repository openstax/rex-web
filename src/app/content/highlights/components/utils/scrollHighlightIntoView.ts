import { Highlight } from '@openstax/highlighter';
import { HTMLElement } from '@openstax/types/lib.dom';
import { scrollIntoView } from '../../../../domUtils';

export default (highlight: Highlight, cardRef?: React.RefObject<HTMLElement>) => {
  const firstElement = highlight.elements[0] as HTMLElement;
  const lastElement = highlight.elements[highlight.elements.length - 1] as HTMLElement;
  const otherElements = [lastElement];
  if (cardRef?.current) {
    otherElements.push(cardRef.current);
  }
  scrollIntoView(firstElement, otherElements);
};
