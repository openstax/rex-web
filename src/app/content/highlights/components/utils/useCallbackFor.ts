import React from 'react';

// Some functions get recreated but shouldn't be reacted to
// This makes a non-changing callback
export default function useCallbackFor<T extends Array<unknown>>(fn: (...args: T) => unknown) {
    const ref = React.useRef(fn);
    const cb = React.useCallback(
        (...args: T) => ref.current(...args),
        []
    );

    ref.current = fn;
    return cb;
}
