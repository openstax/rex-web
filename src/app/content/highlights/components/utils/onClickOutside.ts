import { Element, EventListenerOptions, HTMLElement, MouseEvent } from '@openstax/types/lib.dom';
import curry from 'lodash/fp/curry';
import React from 'react';
import { isHtmlElement } from '../../../../guards';
import { assertDocument, assertWindow } from '../../../../utils';

const targetBelongsToElement = curry((event: MouseEvent, el: React.RefObject<HTMLElement> | HTMLElement): boolean =>
  isHtmlElement(el)
    ? el.contains(event.target as Element)
    : el.current
      ? el.current.contains(event.target as Element)
      : false
);

const onClickOutside = (
  element: React.RefObject<HTMLElement> | Array<React.RefObject<HTMLElement> | HTMLElement>,
  isEnabled: boolean,
  cb: (ev: MouseEvent) => void,
  eventOptions?: EventListenerOptions
) => () => {
  if (typeof document === 'undefined') {
    return () => null;
  }

  const ifOutside = (e: MouseEvent) => {
    if (!(e.target instanceof assertWindow().Element)) {
      return;
    } else if (Array.isArray(element) && element.some(targetBelongsToElement(e))) {
      return;
    } else if (!Array.isArray(element) && targetBelongsToElement(e, element)) {
      return;
    }
    cb(e);
  };

  if (isEnabled) {
    document.addEventListener('click', ifOutside, eventOptions);
  }

  return () => assertDocument().removeEventListener('click', ifOutside, eventOptions);
};

export const useOnClickOutside = (
  element: React.RefObject<HTMLElement> | Array<React.RefObject<HTMLElement> | HTMLElement>,
  isEnabled: boolean,
  cb: (e: MouseEvent) => void,
  eventOptions?: EventListenerOptions
) => {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  React.useEffect(onClickOutside(element, isEnabled, cb, eventOptions), [element, cb, isEnabled]);
};

export default onClickOutside;

const isRefWithHtmlElement = (el: any): el is React.RefObject<HTMLElement> => {
  return el instanceof Object && isHtmlElement(el.current);
};

export const isElementForOnClickOutside = (el: any): el is HTMLElement | React.RefObject<HTMLElement> => {
  return isHtmlElement(el) || isRefWithHtmlElement(el);
};
