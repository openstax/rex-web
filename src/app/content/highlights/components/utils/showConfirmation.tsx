import React from 'react';
import ReactDOM from 'react-dom';
import { RawIntlProvider } from 'react-intl';
import uuid from 'uuid/v4';
import { AppServices } from '../../../../types';
import { assertDocument, assertNotNull } from '../../../../utils';
import ConfirmationModal from '../ConfirmationModal';

export default async(services: AppServices) => {
  const document = assertDocument();
  const domNode = document.createElement('div');

  domNode.id = `dialog-${uuid()}`;
  const root = assertNotNull(document.getElementById('root'), 'root element not found');
  root.insertAdjacentElement('afterend', domNode);

  const userChoice: boolean = await new Promise((resolve) => {
    const confirm = () => resolve(true);
    const deny = () => resolve(false);
    ReactDOM.render(
      <RawIntlProvider value={services.intl}>
        <ConfirmationModal deny={deny} confirm={confirm} />
      </RawIntlProvider>,
      domNode
    );
  });

  ReactDOM.unmountComponentAtNode(domNode);
  assertNotNull(domNode.parentElement, 'parent element not found').removeChild(domNode);

  return userChoice;
};
