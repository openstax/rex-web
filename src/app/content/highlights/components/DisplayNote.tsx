import { Highlight } from '@openstax/highlighter';
import { HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import { useSelector } from 'react-redux';
import styled, { css } from 'styled-components/macro';
import Dropdown, { DropdownItem, DropdownList } from '../../../components/Dropdown';
import Times from '../../../components/Times';
import { textStyle } from '../../../components/Typography/base';
import { useDebouncedWindowSize, useFocusElement } from '../../../reactUtils';
import theme from '../../../theme';
import { mergeRefs } from '../../../utils';
import { highlightStyles } from '../../constants';
import { query } from '../../search/selectors';
import { tocOpen } from '../../selectors';
import { focusHighlight } from '../actions';
import { cardPadding, cardWidth } from '../constants';
import Confirmation from './Confirmation';
import MenuToggle, { MenuIcon } from './MenuToggle';
import TruncatedText from './TruncatedText';
import { isElementForOnClickOutside, useOnClickOutside } from './utils/onClickOutside';

// tslint:disable-next-line:variable-name
const CloseIcon = styled((props) => <Times {...props} aria-hidden='true' focusable='false' />)`
  color: ${theme.color.primary.gray.lighter};
  height: 4.2rem;
  width: 4.2rem;
  padding: 1.6rem;
  display: none;
  position: absolute;
  top: 0;
  right: 0;
  ${theme.breakpoints.touchDeviceQuery(css`
    display: block;
 `)}
`;

export interface DisplayNoteProps {
  highlight: Highlight;
  note: string;
  style: typeof highlightStyles[number];
  isActive: boolean;
  focus: typeof focusHighlight;
  onEdit: () => void;
  onBlur: () => void;
  onRemove: () => void;
  onHeightChange: (ref: React.RefObject<HTMLElement>) => void;
  className: string;
  shouldFocusCard: boolean;
}

// tslint:disable-next-line:variable-name
const DisplayNote = React.forwardRef<HTMLElement, DisplayNoteProps>((
  {note, isActive, highlight, focus, onBlur, onEdit, onRemove, onHeightChange, className, shouldFocusCard},
  ref
) => {
  const [confirmingDelete, setConfirmingDelete] = React.useState<boolean>(false);
  const element = React.useRef<HTMLElement>(null);
  const confirmationRef = React.useRef<HTMLElement>(null);
  const [textToggle, setTextToggle] = React.useState(false);
  const [width] = useDebouncedWindowSize();
  const searchQuery = useSelector(query);
  const isTocOpen = useSelector(tocOpen);

  const onToggle = () => {
    if (!isActive) {
      focus(highlight.id);
    }
  };

  const elements = React.useMemo(
    () => [element, ...highlight.elements].filter(isElementForOnClickOutside),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [element.current, highlight]);

  // Change Event phase so when clicking on another Card,
  // onBlur is called before this Card calls focus.
  useOnClickOutside(elements, isActive, onBlur, { capture: true });

  React.useEffect(() => {
    if (!isActive) {
      setConfirmingDelete(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isActive]);

  React.useEffect(() => {
    const refElement = confirmationRef.current ? confirmationRef : element;
    onHeightChange(refElement);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [element, confirmationRef, confirmingDelete, textToggle, width, isTocOpen, searchQuery]);

  useFocusElement(element, shouldFocusCard);

  return <div className={className} ref={mergeRefs(ref, element)} tabIndex={-1} data-highlight-card>
    <Dropdown toggle={<MenuToggle />} onToggle={onToggle} transparentTab={false}>
      <DropdownList>
        <DropdownItem message='i18n:highlighting:dropdown:edit' onClick={onEdit} />
        <DropdownItem
          message='i18n:highlighting:dropdown:delete'
          data-testid='delete'
          onClick={() => setConfirmingDelete(true)}
        />
      </DropdownList>
    </Dropdown>
    <CloseIcon onClick={onBlur} />
    <label>Note:</label>
    <TruncatedText text={note} isActive={isActive} onChange={() => setTextToggle((state) => !state)} />
    {confirmingDelete && <Confirmation
      ref={confirmationRef}
      data-analytics-label='delete'
      data-analytics-region='confirm-delete-inline-highlight'
      message='i18n:highlighting:confirmation:delete-both'
      confirmMessage='i18n:highlighting:button:delete'
      onConfirm={onRemove}
      onCancel={() => setConfirmingDelete(false)}
    />}
  </div>;
});

export default styled(DisplayNote)`
  width: ${cardWidth}rem;
  overflow: visible;
  background: ${theme.color.neutral.formBackground};
  ${(props: DisplayNoteProps) => props.isActive && css`
    background: ${theme.color.white};
  `}

  > label {
    display: none;
    ${textStyle}
    color: ${(props: DisplayNoteProps) => props.style.focused};
    font-size: 1.4rem;
    line-height: 2rem;
    margin: ${cardPadding * 1.5}rem 0 0 ${cardPadding * 2}rem;
  }

  ${css`
    ${DropdownList}${DropdownList} {
      left: -4rem;
    }
  `}

  ${Dropdown} {
    position: absolute;
    top: 0.8rem;
    right: -0.2rem;

    .focus-within ${MenuIcon} {
      color: ${theme.color.primary.gray.base};
    }

    :focus-within ${MenuIcon} {
      color: ${theme.color.primary.gray.base};
    }
  }

  ${theme.breakpoints.touchDeviceQuery(css`
    width: unset;

    > label {
      display: block;
    }

    ${Dropdown} {
      display: none;
    }
 `)}
`;
