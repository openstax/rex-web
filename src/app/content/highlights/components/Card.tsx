import Highlighter, { Highlight } from '@openstax/highlighter';
import { NewHighlightSourceTypeEnum } from '@openstax/highlighter/dist/api';
import { HTMLElement } from '@openstax/types/lib.dom';
import flow from 'lodash/fp/flow';
import React from 'react';
import { connect, useSelector } from 'react-redux';
import styled from 'styled-components';
import { scrollIntoView } from '../../../domUtils';
import { useFocusIn } from '../../../reactUtils';
import { AppState, Dispatch } from '../../../types';
import { highlightStyles } from '../../constants';
import * as selectHighlights from '../../highlights/selectors';
import * as selectSearch from '../../search/selectors';
import * as selectContent from '../../selectors';
import * as contentSelect from '../../selectors';
import { stripIdVersion } from '../../utils/idUtils';
import {
  clearFocusedHighlight,
  createHighlight,
  focusHighlight,
  requestDeleteHighlight,
  setAnnotationChangesPending,
} from '../actions';
import { HighlightData } from '../types';
import { getHighlightLocationFilterForPage } from '../utils';
import { mainCardStyles } from './cardStyles';
import DisplayNote from './DisplayNote';
import EditCard from './EditCard';
import showConfirmation from './utils/showConfirmation';

export interface CardProps {
  page: ReturnType<typeof selectContent['bookAndPage']>['page'];
  book: ReturnType<typeof selectContent['bookAndPage']>['book'];
  container?: HTMLElement;
  isActive: boolean;
  isTocOpen: boolean;
  hasQuery: boolean;
  highlighter: Highlighter;
  highlight: Highlight;
  create: typeof createHighlight;
  focus: typeof focusHighlight;
  remove: typeof requestDeleteHighlight;
  blur: typeof clearFocusedHighlight;
  setAnnotationChangesPending: typeof setAnnotationChangesPending;
  data?: HighlightData;
  className: string;
  zIndex: number;
  shouldFocusCard: boolean;
  topOffset?: number;
  highlightOffsets?: { top: number, bottom: number };
  onHeightChange: (ref: React.RefObject<HTMLElement>) => void;
}

// tslint:disable-next-line:variable-name
const Card = (props: CardProps) => {
  const annotation = props.data && props.data.annotation;
  const element = React.useRef<HTMLElement>(null);
  const [editing, setEditing] = React.useState<boolean>(!annotation);
  const locationFilters = useSelector(selectHighlights.highlightLocationFilters);
  const hasUnsavedHighlight = useSelector(selectHighlights.hasUnsavedHighlight);

  const { isActive, highlight: { id }, focus } = props;

  const focusCard = React.useCallback(async() => {
    if (!isActive && (!hasUnsavedHighlight || await showConfirmation())) {
      focus(id);
    }
  }, [isActive, hasUnsavedHighlight, id, focus]);

  useFocusIn(element, true, focusCard);

  React.useEffect(() => {
    if (!props.isActive) {
      setEditing(false);
    } else {
      const firstElement = props.highlight.elements[0] as HTMLElement;
      const lastElement = props.highlight.elements[props.highlight.elements.length - 1] as HTMLElement;
      const elements = [lastElement];
      if (element.current) {
        elements.push(element.current);
      }
      scrollIntoView(firstElement, elements);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [element, props.isActive]);

  React.useEffect(() => {
    if (annotation) {
      props.highlight.elements.forEach((el) => (el as HTMLElement).classList.add('has-note'));
    } else {
      props.highlight.elements.forEach((el) => (el as HTMLElement).classList.remove('has-note'));
    }
  }, [props.highlight, annotation]);

  React.useEffect(() => {
    if (!annotation && !props.isActive) {
      props.onHeightChange({ current: null });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [annotation, props.isActive]);

  const location = React.useMemo(() => {
    return props.page && getHighlightLocationFilterForPage(locationFilters, props.page);
  }, [locationFilters, props.page]);

  const locationFilterId = location && stripIdVersion(location.id);

  const { page, book } = props;
  if (!props.highlight.range || !page || !book || !locationFilterId) {
    return null;
  }

  const onRemove = () => {
    if (props.data) {
      props.remove(props.data, {
        locationFilterId,
        pageId: page.id,
      });
    }
  };
  const style = highlightStyles.find((search) => props.data && search.label === props.data.color);

  const onCreate = (isDefaultColor: boolean) => {
    props.create({
      ...props.highlight.serialize().getApiPayload(props.highlighter, props.highlight),
      scopeId: book.id,
      sourceId: page.id,
      sourceMetadata: {bookVersion: book.version},
      sourceType: NewHighlightSourceTypeEnum.OpenstaxPage,
    }, {
      isDefaultColor,
      locationFilterId,
      pageId: page.id,
    });
  };

  const commonProps = {
    className: props.className,
    highlight: props.highlight,
    isActive: props.isActive,
    onBlur: props.blur,
    onHeightChange: props.onHeightChange,
    onRemove,
    ref: element,
    shouldFocusCard: props.shouldFocusCard,
  };

  return <div onClick={focusCard} data-testid='card'>
    {
      !editing && style && annotation ? <DisplayNote
        {...commonProps}
        style={style}
        note={annotation}
        focus={props.focus}
        onEdit={() => setEditing(true)}
      /> : <EditCard
        {...commonProps}
        locationFilterId={locationFilterId}
        hasUnsavedHighlight={hasUnsavedHighlight}
        pageId={page.id}
        onCreate={onCreate}
        setAnnotationChangesPending={props.setAnnotationChangesPending}
        onCancel={() => setEditing(false)}
        data={props.data}
      />
    }
  </div>;
};

// tslint:disable-next-line: variable-name
const StyledCard = styled(Card)`
  ${mainCardStyles}
`;

export default connect(
  (state: AppState, ownProps: {highlight: Highlight}) => ({
    ...selectContent.bookAndPage(state),
    data: selectHighlights.highlights(state).find((search) => search.id === ownProps.highlight.id),
    hasQuery: !!selectSearch.query(state),
    isActive: selectHighlights.focused(state) === ownProps.highlight.id,
    isTocOpen: contentSelect.tocOpen(state),
  }),
  (dispatch: Dispatch) => ({
    blur: flow(clearFocusedHighlight, dispatch),
    create: flow(createHighlight, dispatch),
    focus: flow(focusHighlight, dispatch),
    remove: flow(requestDeleteHighlight, dispatch),
    setAnnotationChangesPending: flow(setAnnotationChangesPending, dispatch),
  })
)(StyledCard);
