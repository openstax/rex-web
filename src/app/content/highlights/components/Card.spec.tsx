import { Highlight } from '@openstax/highlighter';
import { HighlightColorEnum } from '@openstax/highlighter/dist/api';
import { NewHighlightSourceTypeEnum } from '@openstax/highlighter/dist/api';
import { HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import createTestStore from '../../../../test/createTestStore';
import { book, page } from '../../../../test/mocks/archiveLoader';
import createMockHighlight from '../../../../test/mocks/highlight';
import { mockCmsBook } from '../../../../test/mocks/osWebLoader';
import { Store } from '../../../types';
import { assertDocument } from '../../../utils';
import { openToc, receiveBook, receivePage } from '../../actions';
import { highlightStyles } from '../../constants';
import { requestSearch } from '../../search/actions';
import { formatBookData } from '../../utils';
import {
  createHighlight,
  focusHighlight,
  receiveHighlights,
  requestDeleteHighlight,
  setAnnotationChangesPending,
} from '../actions';
import { highlightLocationFilters } from '../selectors';
import { HighlightData } from '../types';
import { getHighlightLocationFilterForPage } from '../utils';
import Card, { CardProps } from './Card';
import DisplayNote from './DisplayNote';
import EditCard from './EditCard';
import showConfirmation from './utils/showConfirmation';

jest.mock('./DisplayNote', () => (jest as any).requireActual('react').forwardRef(
  (props: any, ref: any) => <div ref={ref} mock-display-note {...props} />
));
jest.mock('./EditCard', () => (jest as any).requireActual('react').forwardRef(
  (props: any, ref: any) => <div ref={ref} mock-edit {...props} />
));
jest.mock('./utils/showConfirmation', () => jest.fn(() => new Promise((res) => res(true))));

describe('Card', () => {
  let store: Store;
  let dispatch: jest.SpyInstance;
  let highlight: ReturnType<typeof createMockHighlight>;
  let highlightData: ReturnType<ReturnType<typeof createMockHighlight>['serialize']>['data'];
  let cardProps: Partial<CardProps> & { highlight: Highlight };

  beforeEach(() => {
    store = createTestStore();
    highlight = createMockHighlight('asdf');
    highlightData = highlight.serialize().data;
    dispatch = jest.spyOn(store, 'dispatch');
    highlight.elements = [assertDocument().createElement('span')];
    cardProps = {
      blur: jest.fn(),
      highlight: highlight as unknown as Highlight,
      highlightOffsets: { top: 0, bottom: 0 },
      onHeightChange: () => null,
    };
  });

  it('matches snapshot when focused without note', () => {
    highlight.range.getBoundingClientRect.mockReturnValue({
      bottom: 200,
      top: 100,
    });
    const container = {
      nodeName: 'div',
      nodeType: 1,
      offsetParent: {
        nodeName: 'div',
        nodeType: 1,
        offsetTop: 50,
        tagName: 'DIV',
        title: '',
      },
      tagName: 'DIV',
      title: '',
    } as unknown as HTMLElement;
    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));
    store.dispatch(receivePage({...page, references: []}));
    store.dispatch(receiveHighlights({
      highlights: [
        {
          color: highlightStyles[0].label,
          id: highlightData.id,
        },
      ] as HighlightData[],
      pageId: '123',
    }));
    store.dispatch(focusHighlight(highlight.id));
    const component = renderer.create(<Provider store={store}>
      <Card {...cardProps} container={container} />
    </Provider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot when passed data without note', () => {
    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));
    store.dispatch(receivePage({...page, references: []}));
    store.dispatch(receiveHighlights({
      highlights: [
        {
          id: highlightData.id,
        },
      ] as HighlightData[],
      pageId: '123',
    }));
    store.dispatch(focusHighlight(highlight.id));
    store.dispatch(requestSearch('asdf'));
    const component = renderer.create(<Provider store={store}>
      <Card {...cardProps} />
    </Provider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot without data', () => {
    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));
    store.dispatch(receivePage({...page, references: []}));
    store.dispatch(focusHighlight(highlight.id));
    store.dispatch(openToc()); // added for css coverage
    store.dispatch(requestSearch('asd')); // added for css coverage
    cardProps.highlightOffsets = undefined; // added for css coverage
    const container = assertDocument().createElement('div');
    highlight.range.getBoundingClientRect.mockReturnValue({
      bottom: 200,
      top: 100,
    });
    const component = renderer.create(<Provider store={store}>
      <Card {...cardProps} container={container} />
    </Provider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('unknown style doesn\'t throw', () => {
    store.dispatch(receiveHighlights({
      highlights: [
        {
          color: 'asdfasdfadsf' as HighlightColorEnum,
          id: highlightData.id,
        },
      ] as HighlightData[],
      pageId: '123',
    }));
    expect(() => renderer.create(<Provider store={store}>
      <Card {...cardProps} />
    </Provider>)).not.toThrow();
  });

  it('switches to editing mode when onEdit is triggered', () => {
    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));
    store.dispatch(receivePage({...page, references: []}));
    store.dispatch(receiveHighlights({
      highlights: [
        {
          annotation: 'adsf',
          color: highlightStyles[0].label,
          id: highlightData.id,
        },
      ] as HighlightData[],
      pageId: '123',
    }));

    const component = renderer.create(<Provider store={store}>
      <Card {...cardProps} />
    </Provider>);

    const picker = component.root.findByType(DisplayNote);
    renderer.act(() => {
      picker.props.onEdit();
    });

    expect(() => component.root.findByType(EditCard)).not.toThrow();
  });

  it('switches to display mode when cancelling', () => {
    const data = {
      annotation: 'adsf',
      color: highlightStyles[0].label,
      id: highlightData.id,
    } as HighlightData;
    store.dispatch(receiveHighlights({highlights: [data], pageId: '123'}));
    store.dispatch(focusHighlight(highlight.id));
    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));
    store.dispatch(receivePage({...page, references: []}));

    const component = renderer.create(<Provider store={store}>
      <Card {...cardProps} />
    </Provider>);

    const picker = component.root.findByType(DisplayNote);
    renderer.act(() => {
      picker.props.onEdit();
    });

    const edit = component.root.findByType(EditCard);
    renderer.act(() => {
      edit.props.onCancel();
    });

    expect(() => component.root.findByType(EditCard)).toThrow();
  });

  it('removes when DisplayNote calls onRemove', () => {
    const currentHighlight = {
      annotation: 'adsf',
      color: highlightStyles[0].label,
      id: highlightData.id,
    } as HighlightData;

    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));
    store.dispatch(receivePage({...page, references: []}));
    store.dispatch(receiveHighlights({
      highlights: [currentHighlight],
      pageId: '123',
    }));

    const locationFilters = highlightLocationFilters(store.getState());
    const location = getHighlightLocationFilterForPage(locationFilters, page);
    expect(location).toBeDefined();

    const component = renderer.create(<Provider store={store}>
      <Card {...cardProps} />
    </Provider>);

    const picker = component.root.findByType(DisplayNote);
    renderer.act(() => {
      picker.props.onRemove();
    });

    expect(dispatch).toHaveBeenCalledWith(requestDeleteHighlight(currentHighlight, {
      locationFilterId: location!.id,
      pageId: page.id,
    }));
  });

  it('noops when remove is called but there isn\'t anything to remove', () => {
    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));
    store.dispatch(receivePage({...page, references: []}));
    store.dispatch(focusHighlight(highlight.id));
    const component = renderer.create(<Provider store={store}>
      <Card {...cardProps} />
    </Provider>);

    dispatch.mockClear();

    const picker = component.root.findByProps({ 'mock-edit': true });
    picker.props.onRemove();

    expect(dispatch).not.toHaveBeenCalled();
  });

  it('creates when DisplayNote calls onCreate', () => {
    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));
    store.dispatch(receivePage({...page, references: []}));
    store.dispatch(receiveHighlights({
      highlights: [
        {
          annotation: '',
          color: highlightStyles[0].label,
          id: highlight.id,
        },
      ] as HighlightData[],
      pageId: '123',
    }));
    store.dispatch(focusHighlight(highlight.id));

    dispatch.mockClear();

    const locationFilters = highlightLocationFilters(store.getState());
    const location = getHighlightLocationFilterForPage(locationFilters, page);
    expect(location).toBeDefined();

    const component = renderer.create(<Provider store={store}>
      <Card {...cardProps} />
    </Provider>);

    const editcard = component.root.findByType(EditCard);
    renderer.act(() => {
      editcard.props.onCreate();
    });

    expect(dispatch).toHaveBeenCalledWith(createHighlight({
      ...highlight.serialize().getApiPayload(),
      scopeId: 'testbook1-uuid',
      sourceId: 'testbook1-testpage1-uuid',
      sourceMetadata: {bookVersion: '1.0'},
      sourceType: NewHighlightSourceTypeEnum.OpenstaxPage,
    }, {
      locationFilterId: location!.id,
      pageId: page.id,
    }));
  });

  it('renders null if highlight doen\'t have range', () => {
    (highlight as any).range = undefined;

    const component = renderer.create(<Provider store={store}>
      <Card {...cardProps} />
    </Provider>);

    expect(() => component.root.findByType(EditCard)).toThrow();
  });

  it('renders null if highlight doen\'t have range and its focused', () => {
    (highlight as any).range = undefined;
    store.dispatch(receiveHighlights({
      highlights: [
        {
          color: highlightStyles[0].label,
          id: highlightData.id,
        },
      ] as HighlightData[],
      pageId: '123',
    }));
    store.dispatch(focusHighlight(highlight.id));

    const component = renderer.create(<Provider store={store}>
      <Card {...cardProps} />
    </Provider>);

    expect(() => component.root.findByType(EditCard)).toThrow();
  });

  it('renders null if locationFilter wasn\'t found', () => {
    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));
    store.dispatch(receivePage({...page, id: 'not-in-book', references: []}));

    const component = renderer.create(<Provider store={store}>
      <Card {...cardProps} />
    </Provider>);

    expect(() => component.root.findByType(EditCard)).toThrow();
  });

  it('focuses on click only if it is not already focused', () => {
    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));
    store.dispatch(receivePage({...page, references: []}));
    store.dispatch(receiveHighlights({
      highlights: [
        { id: highlightData.id, annotation: 'asd' },
      ] as HighlightData[],
      pageId: '123',
    }));

    const component = renderer.create(<Provider store={store}>
      <Card {...cardProps} />
    </Provider>);

    expect(dispatch).not.toHaveBeenCalledWith(focusHighlight(highlightData.id));

    const card = component.root.findByProps({ 'data-testid': 'card' });
    renderer.act(() => {
      card.props.onClick();
    });

    expect(dispatch).toHaveBeenCalledWith(focusHighlight(highlightData.id));

    dispatch.mockClear();

    renderer.act(() => {
      card.props.onClick();
    });

    expect(dispatch).not.toHaveBeenCalledWith(focusHighlight(highlightData.id));
  });

  it('displays confirm dialog when there are unsaved changes and user clicks on another card', async() => {
    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));
    store.dispatch(receivePage({...page, references: []}));
    store.dispatch(receiveHighlights({
      highlights: [
        { id: highlightData.id, annotation: 'asd' },
      ] as HighlightData[],
      pageId: '123',
    }));

    store.dispatch(setAnnotationChangesPending(true));

    const component = renderer.create(<Provider store={store}>
      <Card {...cardProps} isActive={false} />
    </Provider>);

    const card = component.root.findByProps({ 'data-testid': 'card' });
    await renderer.act(async() => {
      card.props.onClick();
    });

    expect(showConfirmation).toHaveBeenCalled();
  });
});
