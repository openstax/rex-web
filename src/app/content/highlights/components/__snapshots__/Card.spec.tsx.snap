// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card matches snapshot when focused without note 1`] = `
<div
  data-testid="card"
  onClick={[Function]}
>
  <div
    className="sc-bdVaJa hhNicC"
    data={
      Object {
        "color": "yellow",
        "id": "asdf",
      }
    }
    hasUnsavedHighlight={false}
    highlight={
      Object {
        "addFocusedStyles": [MockFunction],
        "elements": Array [
          <span />,
        ],
        "focus": [MockFunction],
        "getStyle": [MockFunction],
        "id": "asdf",
        "isAttached": [MockFunction],
        "range": Object {
          "commonAncestorContainer": null,
          "getBoundingClientRect": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Object {
                  "bottom": 200,
                  "top": 100,
                },
              },
            ],
          },
        },
        "serialize": [Function],
        "setStyle": [MockFunction],
        "sourceId": "testbook1-testpage1-uuid",
      }
    }
    isActive={true}
    locationFilterId="testbook1-testpage1-uuid"
    mock-edit={true}
    onBlur={[Function]}
    onCancel={[Function]}
    onCreate={[Function]}
    onHeightChange={[Function]}
    onRemove={[Function]}
    pageId="testbook1-testpage1-uuid"
    setAnnotationChangesPending={[Function]}
  />
</div>
`;

exports[`Card matches snapshot when passed data without note 1`] = `
<div
  data-testid="card"
  onClick={[Function]}
>
  <div
    className="sc-bdVaJa iiOeCm"
    data={
      Object {
        "id": "asdf",
      }
    }
    hasUnsavedHighlight={false}
    highlight={
      Object {
        "addFocusedStyles": [MockFunction],
        "elements": Array [
          <span />,
        ],
        "focus": [MockFunction],
        "getStyle": [MockFunction],
        "id": "asdf",
        "isAttached": [MockFunction],
        "range": Object {
          "commonAncestorContainer": null,
          "getBoundingClientRect": [MockFunction],
        },
        "serialize": [Function],
        "setStyle": [MockFunction],
        "sourceId": "testbook1-testpage1-uuid",
      }
    }
    isActive={true}
    locationFilterId="testbook1-testpage1-uuid"
    mock-edit={true}
    onBlur={[Function]}
    onCancel={[Function]}
    onCreate={[Function]}
    onHeightChange={[Function]}
    onRemove={[Function]}
    pageId="testbook1-testpage1-uuid"
    setAnnotationChangesPending={[Function]}
  />
</div>
`;

exports[`Card matches snapshot without data 1`] = `
<div
  data-testid="card"
  onClick={[Function]}
>
  <div
    className="sc-bdVaJa jqlvoB"
    hasUnsavedHighlight={false}
    highlight={
      Object {
        "addFocusedStyles": [MockFunction],
        "elements": Array [
          <span />,
        ],
        "focus": [MockFunction],
        "getStyle": [MockFunction],
        "id": "asdf",
        "isAttached": [MockFunction],
        "range": Object {
          "commonAncestorContainer": null,
          "getBoundingClientRect": [MockFunction] {
            "calls": Array [
              Array [],
              Array [],
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Object {
                  "bottom": 200,
                  "top": 100,
                },
              },
              Object {
                "type": "return",
                "value": Object {
                  "bottom": 200,
                  "top": 100,
                },
              },
              Object {
                "type": "return",
                "value": Object {
                  "bottom": 200,
                  "top": 100,
                },
              },
            ],
          },
        },
        "serialize": [Function],
        "setStyle": [MockFunction],
        "sourceId": "testbook1-testpage1-uuid",
      }
    }
    isActive={true}
    locationFilterId="testbook1-testpage1-uuid"
    mock-edit={true}
    onBlur={[Function]}
    onCancel={[Function]}
    onCreate={[Function]}
    onHeightChange={[Function]}
    onRemove={[Function]}
    pageId="testbook1-testpage1-uuid"
    setAnnotationChangesPending={[Function]}
  />
</div>
`;
